File: src/api/checkServer.ts
import { normalizeServerUrl } from "../storage/persist";
import { fetchSetupStatus, type SetupStatusResponse } from "./auth";
import { ApiError } from "./client";
import { fetchServerHealth } from "./health";

const CHECK_TIMEOUT_MS = 6_000;

export type CheckServerSuccess = {
  ok: true;
  serverUrl: string;
  needsSetup: boolean;
  serverName?: string;
};

export type CheckServerFailure = {
  ok: false;
  message: string;
};

export type CheckServerResult = CheckServerSuccess | CheckServerFailure;

/**
 * Build URL candidates for a manual server address.
 *
 * - Explicit `http(s)://…` keeps that scheme only (matches Android TV / Apple).
 * - Bare host / host:port tries https first, then http.
 */
export function buildManualUrlCandidates(raw: string): string[] {
  const trimmed = raw.trim().replace(/\/+$/, "");
  if (!trimmed) return [];

  if (/^https?:\/\//i.test(trimmed)) {
    const normalized = normalizeServerUrl(trimmed);
    // Normalize scheme casing for stable registry IDs and comparisons.
    const lower = normalized.replace(/^https?:\/\//i, (m) => m.toLowerCase());
    return [lower];
  }

  const withoutSlashes = trimmed.replace(/^\/+/, "");
  return [
    normalizeServerUrl(`https://${withoutSlashes}`),
    normalizeServerUrl(`http://${withoutSlashes}`),
  ];
}

/** User-facing message for transport / TLS / DNS failures (not HTTP error bodies). */
export function networkFailureMessage(err: unknown): string {
  if (err instanceof ApiError) {
    if (err.code === "timeout" || err.status === 408) {
      return "Request timed out. Check the address and try again.";
    }
    return err.message.trim() || "Could not reach a Prairie server at that address.";
  }
  if (err instanceof TypeError) {
    // Browsers (incl. Tizen/webOS) surface TLS / refused / CORS as "Failed to fetch".
    return "Could not reach a Prairie server at that address. Check http vs https and the port.";
  }
  if (err instanceof Error && err.message.trim()) {
    const msg = err.message.trim();
    if (/failed to fetch|networkerror|load failed|network request failed/i.test(msg)) {
      return "Could not reach a Prairie server at that address. Check http vs https and the port.";
    }
    return msg;
  }
  return "Could not reach a Prairie server at that address.";
}

/**
 * Probe a candidate server before showing the login screen.
 * Requires GET /api/v1/auth/setup; health is best-effort for display name.
 */
export async function checkServer(
  serverUrl: string,
  options?: { fetchImpl?: typeof fetch; timeoutMs?: number },
): Promise<CheckServerResult> {
  const normalized = normalizeServerUrl(serverUrl);
  if (!normalized) {
    return { ok: false, message: "Enter a valid Prairie server address." };
  }

  let parsed: URL;
  try {
    parsed = new URL(normalized);
  } catch {
    return { ok: false, message: "Enter a valid Prairie server address." };
  }
  if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
    return { ok: false, message: "Server URL must use http or https." };
  }
  if (parsed.username || parsed.password) {
    return { ok: false, message: "Server URL must not include credentials." };
  }

  const fetchImpl = options?.fetchImpl;
  const timeoutMs = options?.timeoutMs ?? CHECK_TIMEOUT_MS;

  let setup: SetupStatusResponse;
  try {
    setup = await fetchSetupStatus(normalized, fetchImpl, timeoutMs);
  } catch (err) {
    return { ok: false, message: networkFailureMessage(err) };
  }

  let serverName: string | undefined;
  try {
    const health = await fetchServerHealth(normalized, fetchImpl);
    const name = health?.serverName?.trim();
    if (name) serverName = name;
  } catch {
    /* health is optional */
  }

  return {
    ok: true,
    serverUrl: normalized,
    needsSetup: setup.needs_setup === true,
    serverName,
  };
}

/**
 * Try candidates in order until one responds to /auth/setup.
 * Used for manual entry (https → http when scheme omitted).
 */
export async function checkServerCandidates(
  candidates: string[],
  options?: { fetchImpl?: typeof fetch; timeoutMs?: number },
): Promise<CheckServerResult> {
  if (!candidates.length) {
    return { ok: false, message: "Enter a valid Prairie server address." };
  }

  let lastFailure: CheckServerFailure | null = null;
  for (const candidate of candidates) {
    const result = await checkServer(candidate, options);
    if (result.ok) return result;
    lastFailure = result;
  }
  return (
    lastFailure ?? {
      ok: false,
      message: "Could not reach a Prairie server at that address.",
    }
  );
}

File: flutter/packages/prairie_core/lib/src/api/check_server.dart
import 'api_client.dart';
import 'api_error.dart';
import 'auth_api.dart';
import 'health_api.dart';

const _checkTimeout = Duration(seconds: 6);

/// Mirrors `buildManualUrlCandidates` from src/api/checkServer.ts: an
/// explicit scheme keeps only that scheme; a bare host tries https then http.
List<String> buildManualUrlCandidates(String raw) {
  final trimmed = raw.trim().replaceFirst(RegExp(r'/+$'), '');
  if (trimmed.isEmpty) return [];

  if (RegExp(r'^https?://', caseSensitive: false).hasMatch(trimmed)) {
    return [trimmed];
  }

  final withoutSlashes = trimmed.replaceFirst(RegExp(r'^/+'), '');
  return ['https://$withoutSlashes', 'http://$withoutSlashes'];
}

/// User-facing message for transport/TLS/DNS failures (not HTTP error
/// bodies). Mirrors `networkFailureMessage`.
String networkFailureMessage(Object err) {
  if (err is ApiError) {
    if (err.code == 'timeout' || err.status == 408) {
      return 'Request timed out. Check the address and try again.';
    }
    return err.message.trim().isNotEmpty
        ? err.message.trim()
        : 'Could not reach a Prairie server at that address.';
  }
  return 'Could not reach a Prairie server at that address. Check http vs https and the port.';
}

sealed class CheckServerResult {
  const CheckServerResult();
}

class CheckServerSuccess extends CheckServerResult {
  const CheckServerSuccess({required this.serverUrl, required this.needsSetup, this.serverName});
  final String serverUrl;
  final bool needsSetup;
  final String? serverName;
}

class CheckServerFailure extends CheckServerResult {
  const CheckServerFailure(this.message);
  final String message;
}

/// Probes a candidate server before showing the login screen: requires
/// GET /api/v1/auth/setup, with health as best-effort for a display name.
/// Mirrors `checkServer`.
Future<CheckServerResult> checkServer(ApiClient client, String serverUrl) async {
  Uri parsed;
  try {
    parsed = Uri.parse(serverUrl);
  } catch (_) {
    return const CheckServerFailure('Enter a valid Prairie server address.');
  }
  if (parsed.scheme != 'http' && parsed.scheme != 'https') {
    return const CheckServerFailure('Server URL must use http or https.');
  }
  if (parsed.userInfo.isNotEmpty) {
    return const CheckServerFailure('Server URL must not include credentials.');
  }

  SetupStatusResponse setup;
  try {
    setup = await fetchSetupStatus(client, serverUrl, timeout: _checkTimeout);
  } catch (err) {
    return CheckServerFailure(networkFailureMessage(err));
  }

  String? serverName;
  try {
    final health = await fetchServerHealth(client, serverUrl);
    if (health != null && health.serverName.trim().isNotEmpty) serverName = health.serverName.trim();
  } catch (_) {
    // health is optional
  }

  return CheckServerSuccess(serverUrl: serverUrl, needsSetup: setup.needsSetup, serverName: serverName);
}

/// Tries candidates in order until one responds to /auth/setup. Mirrors
/// `checkServerCandidates`.
Future<CheckServerResult> checkServerCandidates(ApiClient client, List<String> candidates) async {
  if (candidates.isEmpty) {
    return const CheckServerFailure('Enter a valid Prairie server address.');
  }
  CheckServerFailure? lastFailure;
  for (final candidate in candidates) {
    final result = await checkServer(client, candidate);
    if (result is CheckServerSuccess) return result;
    lastFailure = result as CheckServerFailure;
  }
  return lastFailure ?? const CheckServerFailure('Could not reach a Prairie server at that address.');
}
